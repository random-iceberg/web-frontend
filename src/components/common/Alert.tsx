import React from "react";

/**
 * Alert variants for styling.
 * @typedef {'success' | 'error' | 'warning' | 'info'} AlertVariant
 */
type AlertVariant = "success" | "error" | "warning" | "info";

/**
 * Props for the Alert component.
 * @typedef {object} AlertProps
 * @property {AlertVariant} [variant='info'] - The style variant of the alert.
 * @property {string} [title] - Optional title for the alert.
 * @property {React.ReactNode} children - The main message content of the alert.
 * @property {string} [className] - Additional CSS classes to apply to the alert container.
 */
type AlertProps = {
  variant?: AlertVariant;
  title?: string; // Optional title
  children: React.ReactNode; // Message content
  className?: string;
};

// Mapping variants to Tailwind classes
const variantStyles: Record<
  AlertVariant,
  { bg: string; border: string; text: string; icon: string; titleText: string }
> = {
  success: {
    bg: "bg-green-50",
    border: "border-green-300",
    text: "text-green-700",
    icon: "✅", // Check Mark Button emoji
    titleText: "text-green-800",
  },
  error: {
    bg: "bg-red-50",
    border: "border-red-300",
    text: "text-red-700",
    icon: "❌", // Cross Mark emoji
    titleText: "text-red-800",
  },
  warning: {
    bg: "bg-yellow-50",
    border: "border-yellow-300",
    text: "text-yellow-700",
    icon: "⚠️", // Warning Sign emoji
    titleText: "text-yellow-800",
  },
  info: {
    bg: "bg-blue-50",
    border: "border-blue-300",
    text: "text-blue-700",
    icon: "ℹ️", // Information Source emoji
    titleText: "text-blue-800",
  },
};

/**
 * Displays messages with different variants (success, error, warning, info).
 * Includes an icon, optional title, and message content.
 *
 * @component
 * @param {AlertProps} props - Props for the Alert component.
 * @example
 * // Error Alert
 * <Alert variant="error" title="Error">
 *   Something went wrong.
 * </Alert>
 *
 * @example
 * // Success Alert without title
 * <Alert variant="success">
 *   Operation completed successfully.
 * </Alert>
 */
const Alert: React.FC<AlertProps> = ({
  variant = "info",
  title,
  children,
  className = "",
}) => {
  const styles = variantStyles[variant];

  return (
    <div
      className={`p-4 rounded-md border ${styles.bg} ${styles.border} ${className}`.trim()}
      role="alert"
    >
      <div className="flex">
        <div className="flex-shrink-0">
          <span aria-hidden="true">{styles.icon}</span>
        </div>
        <div className="ml-3">
          {title && (
            <h3 className={`text-sm font-medium ${styles.titleText}`}>
              {title}
            </h3>
          )}
          <div className={`mt-1 text-sm ${styles.text}`}>{children}</div>
        </div>
      </div>
    </div>
  );
};

export default Alert;
