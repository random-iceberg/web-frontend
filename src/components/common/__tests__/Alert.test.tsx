import React from "react";
import { render } from "@testing-library/react";
import Alert from "../Alert";

describe("Alert Component", () => {
  it("renders info alert correctly", () => {
    const { asFragment } = render(
      <Alert variant="info" title="Information">
        This is an info alert.
      </Alert>,
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("renders success alert correctly", () => {
    const { asFragment } = render(
      <Alert variant="success" title="Success!">
        Operation successful.
      </Alert>,
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("renders error alert correctly", () => {
    const { asFragment } = render(
      <Alert variant="error" title="Error Occurred">
        Something went wrong.
      </Alert>,
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("renders warning alert correctly", () => {
    const { asFragment } = render(
      <Alert variant="warning" title="Warning">
        Please be careful.
      </Alert>,
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("renders alert without title correctly", () => {
    const { asFragment } = render(
      <Alert variant="info">This is an info alert without a title.</Alert>,
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("renders with additional className", () => {
    const { asFragment } = render(
      <Alert variant="success" className="my-custom-class">
        Custom class test.
      </Alert>,
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
