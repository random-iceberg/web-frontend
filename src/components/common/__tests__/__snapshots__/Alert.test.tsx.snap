// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Alert Component renders alert without title correctly 1`] = `
<DocumentFragment>
  <div
    class="p-4 rounded-md border bg-blue-50 border-blue-300"
    role="alert"
  >
    <div
      class="flex"
    >
      <div
        class="flex-shrink-0"
      >
        <span
          aria-hidden="true"
        >
          ℹ️
        </span>
      </div>
      <div
        class="ml-3"
      >
        <div
          class="mt-1 text-sm text-blue-700"
        >
          This is an info alert without a title.
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Alert Component renders error alert correctly 1`] = `
<DocumentFragment>
  <div
    class="p-4 rounded-md border bg-red-50 border-red-300"
    role="alert"
  >
    <div
      class="flex"
    >
      <div
        class="flex-shrink-0"
      >
        <span
          aria-hidden="true"
        >
          ❌
        </span>
      </div>
      <div
        class="ml-3"
      >
        <h3
          class="text-sm font-medium text-red-800"
        >
          Error Occurred
        </h3>
        <div
          class="mt-1 text-sm text-red-700"
        >
          Something went wrong.
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Alert Component renders info alert correctly 1`] = `
<DocumentFragment>
  <div
    class="p-4 rounded-md border bg-blue-50 border-blue-300"
    role="alert"
  >
    <div
      class="flex"
    >
      <div
        class="flex-shrink-0"
      >
        <span
          aria-hidden="true"
        >
          ℹ️
        </span>
      </div>
      <div
        class="ml-3"
      >
        <h3
          class="text-sm font-medium text-blue-800"
        >
          Information
        </h3>
        <div
          class="mt-1 text-sm text-blue-700"
        >
          This is an info alert.
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Alert Component renders success alert correctly 1`] = `
<DocumentFragment>
  <div
    class="p-4 rounded-md border bg-green-50 border-green-300"
    role="alert"
  >
    <div
      class="flex"
    >
      <div
        class="flex-shrink-0"
      >
        <span
          aria-hidden="true"
        >
          ✅
        </span>
      </div>
      <div
        class="ml-3"
      >
        <h3
          class="text-sm font-medium text-green-800"
        >
          Success!
        </h3>
        <div
          class="mt-1 text-sm text-green-700"
        >
          Operation successful.
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Alert Component renders warning alert correctly 1`] = `
<DocumentFragment>
  <div
    class="p-4 rounded-md border bg-yellow-50 border-yellow-300"
    role="alert"
  >
    <div
      class="flex"
    >
      <div
        class="flex-shrink-0"
      >
        <span
          aria-hidden="true"
        >
          ⚠️
        </span>
      </div>
      <div
        class="ml-3"
      >
        <h3
          class="text-sm font-medium text-yellow-800"
        >
          Warning
        </h3>
        <div
          class="mt-1 text-sm text-yellow-700"
        >
          Please be careful.
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Alert Component renders with additional className 1`] = `
<DocumentFragment>
  <div
    class="p-4 rounded-md border bg-green-50 border-green-300 my-custom-class"
    role="alert"
  >
    <div
      class="flex"
    >
      <div
        class="flex-shrink-0"
      >
        <span
          aria-hidden="true"
        >
          ✅
        </span>
      </div>
      <div
        class="ml-3"
      >
        <div
          class="mt-1 text-sm text-green-700"
        >
          Custom class test.
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
